### PR TITLE
[eos-noon] implement dice contract

### PR DESCRIPTION
**Pull request by elmato**
* Add find_table/get_index functions to tester
* abi_generator: Implement option to simplify one field structs
* abi_generator: Improve table key detection
* Fix native_tokens type declaration
* Implement dice contract
* Fix object_key_value
* eoslib: Implement missing functions in table::secondary_index
* Copy eosio.system to installation folder
* Add tests for dice contract

### CODE REVIEW

File: contracts/eoslib/db.hpp [hunk 1]
```diff
@@ -248,8 +248,9 @@ struct table {
        *
        *  @return true if successful read.
        */
-      static bool lower_bound( const PrimaryType& p, Record& r ) {
-         return impl::lower_bound_primary( scope, code, table_n, &p &r, sizeof(Record) ) == sizeof(Record);
+      static bool lower_bound( const PrimaryType& p, Record& r, uint64_t s = scope ) {
+         *reinterpret_cast<PrimaryType*>(&r) = p;
```
Comment (SheldonHH) [inline contracts/eoslib/db.hpp]: `*reinterpret_cast` does not compile to any CPU instructions.

File: contracts/eosio.system/eosio.system.hpp [hunk 2]
```diff
@@ -19,7 +20,7 @@ namespace eosiosystem {
    /**
    * We create the native EOSIO token type
    */
-   typedef eosio::token<uint64_t,N(eosio)> native_tokens;
+   typedef eosio::token<N(eosio), S(4,EOS), uint64_t> native_tokens;
```
Comment (SheldonHH) [inline contracts/eosio.system/eosio.system.hpp]: What's the meaning of `S(4,EOS)`, where is `S()` defined?

File: contracts/CMakeLists.txt [hunk 3]
```diff
@@ -9,6 +9,8 @@ add_subdirectory(test_api)
 #add_subdirectory(simpledb)
 #add_subdirectory(storage)
 #add_subdirectory(social)
+add_subdirectory(dice)
 
 install( DIRECTORY eoslib DESTINATION include/ )
+install( DIRECTORY eosio.system DESTINATION include/ )
```
Comment (SheldonHH) [inline contracts/CMakeLists.txt]: two struct is used for eosio.system

File: contracts/eosio.system/eosio.system.hpp (hunk 2 above)
Comment (SheldonHH) [reply contracts/eosio.system/eosio.system.hpp]: Answer to myself: 
`S()` macro is defined in`#define S(P,X) ::eosio::string_to_symbol(P,#X)` 
- eosio.system.hpp has `#included <token.hpp>`, 
- token.hpp has `#include<assets.hpp>`

**Comment by ukarlsson** [comment]
I'd like to try out this contract, any chance that you could rebase it? :)

**Comment by elmato** [comment]
replaced by #1663

**Pull request closed without merging**